inner_products/src/lib.rs

```diff
@@ -0,0 +1,165 @@
+use algebra::{
+    bytes::ToBytes,
+    curves::PairingEngine,
+    fields::{Field, PrimeField},
+    groups::Group,
+};
+use std::{
+    error::Error as ErrorTrait,
+    fmt::{Display, Formatter, Result as FmtResult},
+    io::{Result as IoResult, Write},
+    marker::PhantomData,
+    ops::{Add, Mul, MulAssign},
+};
+
+pub type Error = Box<dyn ErrorTrait>;
+
+#[derive(Debug)]
+pub enum InnerProductError {
+    MessageLengthInvalid(usize, usize),
+}
+
+impl ErrorTrait for InnerProductError {
+    fn source(self: &Self) -> Option<&(dyn ErrorTrait + 'static)> {
+        None
+    }
+}
+
+impl Display for InnerProductError {
+    fn fmt(self: &Self, f: &mut Formatter<'_>) -> FmtResult {
+        let msg = match self {
+            InnerProductError::MessageLengthInvalid(left, right) => {
+                format!("left length, right length: {}, {}", left, right)
+            }
+        };
+        write!(f, "{}", msg)
+    }
+}
+
+pub trait InnerProduct: Copy {
+    type LeftMessage;
+    type RightMessage;
+    type Output;
+
+    fn inner_product(
+        left: &[Self::LeftMessage],
+        right: &[Self::RightMessage],
+    ) -> Result<Self::Output, Error>;
+}
+
+#[derive(Copy, Clone)]
+pub struct PairingInnerProduct<P: PairingEngine> {
+    _pair: PhantomData<P>,
+}
+
+impl<P: PairingEngine> InnerProduct for PairingInnerProduct<P> {
+    type LeftMessage = P::G1Projective;
+    type RightMessage = P::G2Projective;
+    type Output = ExtensionFieldElement<P>;
+
+    fn inner_product(
+        left: &[Self::LeftMessage],
+        right: &[Self::RightMessage],
+    ) -> Result<Self::Output, Error> {
+        if left.len() != right.len() {
+            return Err(Box::new(InnerProductError::MessageLengthInvalid(
+                left.len(),
+                right.len(),
+            )));
+        };
+        Ok(ExtensionFieldElement(
+            left.iter()
+                .zip(right)
+                .map(|(v, a)| P::pairing(v.clone().into(), a.clone().into()))
+                .product(),
```

Comment:
This does n final exponentations, you probably want to use something like this: https://github.com/scipr-lab/ripp/blob/e18c63cdda5710c9c65a7ec43532ac8281af3f35/src/lib.rs#L150

Reply:
:+1: 

Done. I used:

https://github.com/scipr-lab/zexe/blob/master/algebra-core/src/curves/mod.rs#L73

I think adding in parallelization support can be done in a future pass?